### PR TITLE
Add calculator implementation with FP

### PR DESCRIPTION
**Pull request by yani-valeva**

### CODE REVIEW

File: assets/css/style.css [hunk 1]
```diff
@@ -0,0 +1,77 @@
+.module, body .wrapper .all-added-nums-sum .result, body .wrapper .all-added-nums-sum .added-nums-container .current-added-elem {
+  color: #625F5F;
```
Comment (aleksandrenko) [inline assets/css/style.css:2]: The products and map file of the sass should not be committed. 
You can use .gitignore for this

File: index.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta http-equiv="X-UA-Compatible" content="ie=edge">
+    <title>Calculator</title>
+    <link rel="stylesheet" href="./assets/css/style.css">
+</head>
+<body>
+    <div class="wrapper">
```
Comment (aleksandrenko) [inline index.html:13]: The end style of the page is lacking some styling from the provided design.

File: js/sum.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+let addBtn = document.getElementById('add-btn');
```
Comment (aleksandrenko) [inline js/sum.js]: What will happen if we move the script tag in the body in the header? 
Is this going to work?

Comment (aleksandrenko) [inline js/sum.js]: - This implementation is lacking "functional programming" common approaches. 
Its easier to call it procedural.

File: js/sum.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+let newNum = document.getElementById('new-num');
+let numsContainer = document.getElementById('added-nums-container');
```
Comment (aleksandrenko) [inline js/sum.js]: Are you sure those elements with searched Ids are existing at this point?

File: js/sum.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+let sumElem = document.getElementById('sum');
+let isOdd = false;
```
Comment (aleksandrenko) [inline js/sum.js]: Is this used to just style the elements? 
This can be achieved with css only and no js logic. 
Please refactor.

File: js/sum.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+addBtn.addEventListener('click', () => {
```
Comment (aleksandrenko) [inline js/sum.js]: This function contains side effects.

File: js/sum.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    if (newNum.value === '') {
+        alert('Enter a number, please!');
+        return;
+    }
+
+    isOdd = !isOdd;
+    let currentElem = document.createElement('div');
```
Comment (aleksandrenko) [inline js/sum.js]: This approach works but it's not very optimal and readable. It can be done with a string template literal. 
Also creating elements directly on existing dom elements and changing attributes may trigger browser rerendering and cause performance issues. This can be worked around with a dom fragment.

File: js/sum.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    isOdd = !isOdd;
+    let currentElem = document.createElement('div');   
+    isOdd ?  currentElem.setAttribute('class', 'current-added-elem odd') : currentElem.setAttribute('class', 'current-added-elem');  
+    let addedNum = document.createElement('span');
+    let deleteBtn = document.createElement('button');
+    deleteBtn.style.all = 'unset';
+    deleteBtn.innerText = '(delete)';
+    deleteBtn.setAttribute('id', 'delete');
+    deleteBtn.setAttribute('class', 'delete-btn'); 
+    let textNode = document.createTextNode((+newNum.value).toFixed(2));
```
Comment (aleksandrenko) [inline js/sum.js]: It's not a good place to change/cut the data. The data should be limited/validated before this point of business logic.

File: js/sum.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    addedNum.appendChild(textNode);
+    currentElem.appendChild(addedNum)
+    currentElem.appendChild(deleteBtn);
+    numsContainer.appendChild(currentElem);
+    sumElem.innerText = (+sumElem.innerText + +addedNum.innerText).toFixed(2);
+
+    deleteBtn.addEventListener('click', (event) => {
```
Comment (aleksandrenko) [inline js/sum.js]: Adding an event listener for each newly created button can cause performance issues. If we have 1000 button we will have 1000 listeners with mostly duplicated logic. Can we optimize this?

File: js/sum.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        const currentDelBtn = event.target;
+        sumElem.innerText = (+sumElem.innerText) - (+currentDelBtn.previousSibling.innerText);
```
Comment (aleksandrenko) [inline js/sum.js]: There is too much magic here. A string is coming and it's converted into a number.

File: js/sum.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    sumElem.innerText = (+sumElem.innerText + +addedNum.innerText).toFixed(2);
+
+    deleteBtn.addEventListener('click', (event) => {
+        const currentDelBtn = event.target;
+        sumElem.innerText = (+sumElem.innerText) - (+currentDelBtn.previousSibling.innerText);   
+        currentDelBtn.parentNode.parentNode.removeChild(currentDelBtn.parentNode);
+        let elements = numsContainer.getElementsByTagName('div');
+        for (let i = 0; i < elements.length; i++) {
+            elements[i].removeAttribute('class');
+            i % 2 === 0 ? elements[i].setAttribute('class', 'current-added-elem odd') : elements[i].setAttribute('class', 'current-added-elem');
```
Comment (aleksandrenko) [inline js/sum.js]: This 'odd' element logic can be done in css.

File: src/scss/style.scss [hunk 12]
```diff
@@ -0,0 +1,92 @@
+.module {
+    color: #625F5F;
+    border-radius: 2px;
+    padding: 15px;
+}
+
+body {
```
Comment (aleksandrenko) [inline src/scss/style.scss:7]: It is a good practice to place the common css styles on top. 
Also, a well-proven practice is to use styling like normalize or reset css.

File: src/scss/style.scss [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    font-family: Arial, Helvetica, sans-serif;
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    color: #C8C4C3;
+    min-height: 100vh;
+
+    .wrapper {
```
Comment (aleksandrenko) [inline src/scss/style.scss:19]: Adding this selector as children will generate extra long selector. Is there any value to add extra "body" in front of every selector from here to the end?

File: src/scss/style.scss [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        display: flex;
+        border: 1px solid #C8C4C3;
+        border-radius: 10px;
+        justify-content: center;
+        flex-direction: column;
+        -webkit-box-shadow: 6px 9px 5px -4px rgba(0,0,0,0.44);
```
Comment (aleksandrenko) [inline src/scss/style.scss]: You are using sass, this can be extracted into a mixin or a function. It's not a big deal but still, it's a way to clear the code.

File: src/scss/style.scss [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+            display: flex;
+        }
+    
+        
+        .all-added-nums-sum {
+            border-right: 1px solid #C8C4C3;
+            padding: 10px;
+
+            .added-nums-container {
+                height: 60vh;
```
Comment (aleksandrenko) [inline src/scss/style.scss:46]: In this layout setup, the size of the whole panel is defined from its content.  It should be the other way around.
You should have a defined (somehow) size of the panel and the content should take its space.

File: js/sum.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
+let addBtn = document.getElementById('add-btn');
+let newNum = document.getElementById('new-num');
+let numsContainer = document.getElementById('added-nums-container');
+let sumElem = document.getElementById('sum');
+let isOdd = false;
+
+addBtn.addEventListener('click', () => {
+    if (newNum.value === '') {
```
Comment (aleksandrenko) [inline js/sum.js]: You are counting on the ui for the validation. This is not a good practice. And it's not very obvious. If some reason the input's type will be changed to 'text' this validation will not work and let strings trough.
And also I still can write something like "555e" into the input itself.

File: src/scss/style.scss [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+                @extend .module;
+            }
+        }
+    
+        .new-num-part {
+            display: flex;
+            flex-direction: column;
+            padding: 10px;
+    
+            * {
```
Comment (aleksandrenko) [inline src/scss/style.scss:78]: Using '*' can lead to slow browser work. Try to style your elements in any other way if there is any.

File: src/scss/style.scss (hunk 12 above)
Comment (aleksandrenko) [inline src/scss/style.scss:7]: Is there any font-size styling?

**Review by aleksandrenko — COMMENTED** [review]
#1

File: js/sum.js (hunk 3 above)
Comment (yani-valeva) [reply js/sum.js]: I agree. I fixed it using window eventListener('DOMContentLoaded').

File: index.html (hunk 2 above)
Comment (yani-valeva) [reply index.html:13]: Can you explain in more detail what you mean?

File: js/sum.js [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+        currentElem.appendChild(deleteBtn);
+        numsContainer.appendChild(currentElem);
+        sumElem.innerText = (+sumElem.innerText + +addedNum.innerText).toFixed(2);
+
+        deleteBtn.addEventListener('click', (event) => {
+            const currentDelBtn = event.target;
+            sumElem.innerText = (+sumElem.innerText) - (+currentDelBtn.previousSibling.innerText);
+            currentDelBtn.parentNode.parentNode.removeChild(currentDelBtn.parentNode);
+            let elements = numsContainer.getElementsByTagName('div');
+            for (let i = 0; i < elements.length; i++) {
```
Comment (VladimirMilov) [inline js/sum.js:35]: You can achive the 'stripe effect' only with css you dont need to add 'odd' class here.
Take a look on the :nth-child selector. It accepts even and odd as well! :)
In that way you will be able to remove a LOT of code from here.

File: js/sum.js [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+                elements[i].removeAttribute('class');
+                i % 2 === 0 ? elements[i].setAttribute('class', 'current-added-elem odd') : elements[i].setAttribute('class', 'current-added-elem');
+            }
+
+            isOdd = elements.length % 2 === 0 ? false : true;
+        });
+    });
+
+    newNum.addEventListener('click', () => {
```
Comment (VladimirMilov) [inline js/sum.js]: Let's do this on .add-btn click. Its a bit more intuitive

File: js/sum.js [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        newNum.value = '';
```
Comment (VladimirMilov) [inline js/sum.js:45]: Let's do this on .add-btn click. Its a bit more intuitive

File: js/sum.js [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+        isOdd = !isOdd;
+        let currentElem = document.createElement('div');
+        isOdd ? currentElem.setAttribute('class', 'current-added-elem odd') : currentElem.setAttribute('class', 'current-added-elem');
+        let addedNum = document.createElement('span');
+        let deleteBtn = document.createElement('button');
+        deleteBtn.style.all = 'unset';
+        deleteBtn.innerText = '(delete)';
+        deleteBtn.setAttribute('id', 'delete');
+        deleteBtn.setAttribute('class', 'delete-btn');
```
Comment (VladimirMilov) [inline js/sum.js:22]: lets try to be a bit more 'functional'. We can extract the code above in a separate function that will create a new dom element(without appending it), add some text and attributes and return this element.

File: js/sum.js [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+            alert('Enter a number, please!');
+            return;
+        }
+
+        isOdd = !isOdd;
+        let currentElem = document.createElement('div');
+        isOdd ? currentElem.setAttribute('class', 'current-added-elem odd') : currentElem.setAttribute('class', 'current-added-elem');
+        let addedNum = document.createElement('span');
+        let deleteBtn = document.createElement('button');
+        deleteBtn.style.all = 'unset';
```
Comment (VladimirMilov) [inline js/sum.js:19]: What is the reason for doing this?

File: js/sum.js [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+        deleteBtn.setAttribute('id', 'delete');
+        deleteBtn.setAttribute('class', 'delete-btn');
+        let textNode = document.createTextNode((+newNum.value).toFixed(2));
+        addedNum.appendChild(textNode);
+        currentElem.appendChild(addedNum)
+        currentElem.appendChild(deleteBtn);
+        numsContainer.appendChild(currentElem);
+        sumElem.innerText = (+sumElem.innerText + +addedNum.innerText).toFixed(2);
+
+        deleteBtn.addEventListener('click', (event) => {
```
Comment (VladimirMilov) [inline js/sum.js:30]: As Niki pointed in the previous commit this can be improved.

File: .vscode/settings.json [hunk 24]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (VladimirMilov) [inline .vscode/settings.json]: This file should be in .gitignore as well.

File: index.html [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+        <h6>Adding numbers is hard!</h6>
+        <div class="content">
+            <section class="all-added-nums-sum">
+                <div id="added-nums-container" class="added-nums-container"></div>
+                <div id="sum" class="result">0.00</div>
+            </section>
+            <aside class="new-num-part">
+                <input required id="new-num" class="new-num" type="number" name="num" step=".01" placeholder="0.00">
+                <button id="add-btn" class="add-btn">Add number</button>
```
Comment (VladimirMilov) [inline index.html:22]: is there a reason for duplicating both class and ids in the html elements?

File: src/scss/style.scss [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+            border-right: 1px solid #C8C4C3;
+            padding: 10px;
+
+            .added-nums-container {
+                height: 60vh;
+                overflow: auto;
+                width: 30vw;
+
+                .current-added-elem {
+                    @extend .module;
```
Comment (VladimirMilov) [inline src/scss/style.scss:51]: Awesome 👍

File: js/sum.js (hunk 22 above)
Comment (yani-valeva) [reply js/sum.js:19]: Тhe reason is that I want odd elements to have an extra class to add background. I create the children of this element and remove the button's default style because it does not look like a button. What would you recommend as an alternative?

File: index.html (hunk 25 above)
Comment (yani-valeva) [reply index.html:22]: Тhe reason I do this is that as far as I know, it's nice to use the ids for the js files and the classes for the styling. :)

File: js/sum.js (hunk 22 above)
Comment (VladimirMilov) [reply js/sum.js:19]: okay so you dont need this one: 'I want odd elements to have an extra class to add background.'
As I mentioned in one of my other comments you can achieve this effect only with css.
Take a look at :nth-child(odd).
Removing styles and adding additional ones on the delete button can again be done only with css. In that way you will clear a bit your code and make it a bit more readable.

File: index.html (hunk 2 above)
Comment (aleksandrenko) [reply index.html:13]: - the font family and size are not like in the design
- the body bg color is not like in the design
- the add button color is not the same as the one in the design
- add/even bg color is not like in the design
- the wrapper shadow is not like in the design

You can https://www.figma.com/file/GPLBZNBQqu4WlRvoB9kbZt9c/Untitled?node-id=1%3A3 open the figma design and select the most right tab "Code" in the right panel. when selecting elements from the desing there you can see suggested css for each.

File: index.html (hunk 25 above)
Comment (aleksandrenko) [reply index.html:22]: Agree with Yanitza. Classes are for styling, ids for getting a reference to the dom element.
In modern apps (react, ang..) ids are not used.

File: js/sum.js (hunk 6 above)
Comment (nikoletavnv) [reply js/sum.js]: You can specify what kind of side effects you see

File: js/sum.js [hunk 27]
```diff
@@ -0,0 +1,47 @@
+window.addEventListener('DOMContentLoaded', () => {
+    let addBtn = document.getElementById('add-btn');
+    let newNum = document.getElementById('new-num');
+    let numsContainer = document.getElementById('added-nums-container');
+    let sumElem = document.getElementById('sum');
+    let isOdd = false;
+
+    addBtn.addEventListener('click', () => {
```
Comment (nikoletavnv) [inline js/sum.js:8]: This function is very long and is doing a lot of things at the same time: styling, constructing and calculating. You can try to separate it to smaller functions with clear responsibility. Basically the whole logic is happening in 1 method and it is hard to follow the logic there.

File: index.html (hunk 2 above)
Comment (yani-valeva) [reply index.html:13]: Thank you! I think that I correct it. :)

File: js/sum.js (hunk 16 above)
Comment (yani-valeva) [reply js/sum.js]: Fixed :)

File: src/scss/style.scss (hunk 12 above)
Comment (yani-valeva) [reply src/scss/style.scss:7]: Fixed

**Comment by aleksandrenko** [comment]
My oop implementation:
https://github.com/aleksandrenko/fop/pull/1
